### ISSUE TITLE
Add more charts

### ISSUE THREAD

**Issue by nesnoj** [rl-institut/WAM_APP_stemp_abw#22]
I'd like some more chart types in the tool. Could you please take care of this @4lm?

The most important is a pie chart, I'll add more types later to this issue.
Some time ago I created a simple skeleton, I pushed it to a new feature branch where you can continue if you like to:
https://github.com/rl-institut/WAM_APP_stemp_abw/tree/feature/add_more_charts

The chart type demos incl. code can be found [here](https://www.highcharts.com/demo), the [PieChart is here](https://www.highcharts.com/demo/pie-basic).

**Comment by 4lm** [comment]
Hi @nesnoj, as spoken via Email I will take care of adding more charts. The first will be the PieChart. For doing so, I will continue to work on your feature branch: https://github.com/rl-institut/WAM_APP_stemp_abw/tree/feature/add_more_charts

**Comment by nesnoj** [comment]
Ok, here comes my wishlist.
As it is quite extensive, therefore I've put priorities beside (3=high...1=low).

- [Timeseries zoomable](https://www.highcharts.com/demo/line-time-series) (3)
- [Stacked area](https://www.highcharts.com/demo/area-stacked) (3)
- [Percentage area](https://www.highcharts.com/demo/area-stacked-percent) (2)
- [Stacked column](https://www.highcharts.com/demo/column-stacked) (3)
- [Stacked and grouped column](https://www.highcharts.com/demo/column-stacked-and-grouped) (3)
- [Bubble chart](https://www.highcharts.com/demo/bubble) (1)
- [Scatter](https://www.highcharts.com/demo/scatter) (1)
- [Synchronized charts](https://www.highcharts.com/demo/synchronized-charts) (1)
- [Solid gauge](https://www.highcharts.com/demo/gauge-solid) (2)
- [Sankey diagram](https://www.highcharts.com/demo/sankey-diagram) (1)
- [Heatmap](https://www.highcharts.com/demo/heatmap-canvas) (1)

Could you drop a note as soon as you'll have finished top prio (3) so we can see how much time is left for this?
Merci!

**Comment by nesnoj** [comment]
As written in the mail today, I made some adjustments to WAM's highcharts:
https://github.com/rl-institut/WAM/pull/19

I changed our [branch](https://github.com/rl-institut/WAM_APP_stemp_abw/tree/feature/add_more_charts) accordingly, added pie charts in the popups & results and fixed the options.

It was quite a pain to adjust the data format to be in accordance to the expected depending on the chart type. Especially for the pie chart it seems to be different, the format has to look like the one in the[ example of the python HC package](https://github.com/kyper-data/python-highcharts/blob/e08d4ff81311d2fef8fb6ec61650b1198ed2e346/examples/ipynb/highcharts/pie-basic.ipynb). For the pie I did not change the data processing in the WAM but the preprocessing in our package to return a dict.

One important thing is to fully refactor the result class structure `ResultAnalysisVisualization` as I've quickly added it last year to show sample graphs. Let's sort this out on Tuesday!

**Comment by nesnoj** [comment]
Here comes the first 3 charts I'd like to have in the popup per layer:

**Layer: reg_mun_pop**
- description: Bevölkerungentwicklung (Prognose)
- type: line
- unit: capita
- data:
  - MunData.pop_2017
  - MunData.pop_2030
  - MunData.pop_2050

**Layer: reg_mun_energy_re_el_dem_share**
- description: Anteil Energie aus EE an Strombedarf nach Quelle
- type: pie
- unit: %
- data:
  - MunData.gen_el_energy_wind / RegMunDemElEnergy.dem_el_energy
  - MunData.gen_el_energy_pv_roof / RegMunDemElEnergy.dem_el_energy
  - MunData.gen_el_energy_pv_ground / RegMunDemElEnergy.dem_el_energy
  - MunData.gen_el_energy_hydro / RegMunDemElEnergy.dem_el_energy

**Layer: reg_mun_gen_energy_re**
- description: Gewonnene Energie aus EE nach Quelle
- type: pie
- unit: GWh and %
- data:
  - MunData.gen_el_energy_wind
  - MunData.gen_el_energy_pv_roof
  - MunData.gen_el_energy_pv_ground
  - MunData.gen_el_energy_hydro

**===== EDIT 21.03.19 =====**
Some more popup charts:

** Layer: reg_mun_gen_energy_re_per_capita**
- description: Gewonnene Energie aus EE je EinwohnerIn nach Quelle
- type: pie
- unit: MWh and %
- data: analog to reg_mun_gen_energy_re

** Layer: reg_mun_gen_energy_re_density**
- description: Gewonnene Energie aus EE je km² nach Quelle
- type: pie
- unit: MWh and %
- data: analog to reg_mun_gen_energy_re

** Layer: reg_mun_gen_cap_re**
- description: Installierte Leistung nach Quelle
- type: pie
- unit: MW and %
- data:
  - MunData.gen_capacity_wind / RegMunGenCapRe.gen_cap_re
  - MunData.gen_capacity_pv_roof_large / RegMunGenCapRe.gen_cap_re
  - MunData.gen_capacity_pv_ground / RegMunGenCapRe.gen_cap_re
  - MunData.gen_capacity_hydro / RegMunGenCapRe.gen_cap_re
  - MunData.gen_capacity_bio / RegMunGenCapRe.gen_cap_re

** Layer: reg_mun_gen_cap_re_density**
- description: Installierte Leistung je km² nach Quelle
- type: pie
- unit: MW and %
- data:
  - MunData.gen_capacity_wind / MunData.area
  - MunData.gen_capacity_pv_roof_large / MunData.area
  - MunData.gen_capacity_pv_ground / MunData.area
  - MunData.gen_capacity_hydro / MunData.area
  - MunData.gen_capacity_bio / MunData.area

** Layer: reg_mun_dem_el_energy**
- description: Strombedarf nach Verbrauchergruppe
- type: pie
- unit: GWh and %
- data:
  - MunData.dem_el_energy_hh (Haushalte)
  - MunData.dem_el_energy_rca (GHD und Landw.)
  - MunData.dem_el_energy_ind (Industrie)

** Layer: reg_mun_dem_el_energy_per_capita**
- description: Strombedarf je EinwohnerIn nach Verbrauchergruppe
- type: pie
- unit: kWh and %
- data: analog to reg_mun_dem_el_energy

** Layer: reg_mun_dem_th_energy**
- description: Wärmebedarf nach Verbrauchergruppe
- type: pie
- unit: GWh and %
- data:
  - MunData.dem_th_energy_hh  (Haushalte)
  - MunData.dem_th_energy_rca (GHD und Landw.)

** Layer: reg_mun_dem_th_energy_per_capita**
- description: Wärmebedarf je EinwohnerIn nach Verbrauchergruppe
- type: pie
- unit: kWh and %
- data: analog to reg_mun_dem_th_energy

**Comment by 4lm** [comment]
Update, @nesnoj and I introduced prio 4 for the creation of the highcharts types: 

Stacked column (4)
Stacked and grouped column (4)
Solid gauge (4)

**Comment by 4lm** [comment]
Hi @nesnoj,

as spoken on the phone a few minutes ago, I pushed some charts labels data and integrated an example for three stacked, grouped, column charts. So that you can give it some love and make our charts groupable via stack param :)

- It's on branch feature/add_more_charts: https://github.com/rl-institut/WAM_APP_stemp_abw/tree/feature/add_more_charts

- Depends on following commits:
-- Add chart data and integrate example for stacked, grouped, column chart: e6c33d366b3f56830c0b538d7819674f9a8222d6
-- Make add_pandas_data_set take kwargs: 2dee8882dd82b45266853ac160a503db3eae6fa7

- Example is integrated in results section: http://127.0.0.1:8000/stemp_abw/app/#results-panel4

- python highcharts howto example concerning the topic (stacked-grouped-column chart), which mirrors 1:1 the JS only example on highcharts.com: https://github.com/kyper-data/python-highcharts/blob/master/examples/highcharts/column-stacked-and-grouped.py

**Comment by 4lm** [comment]
I will now try to integrate solid gauge charts.

**Comment by 4lm** [comment]
Hi @nesnoj,

TL;DR: Looks like python-highcharts, the wrapper around highcharts.js we are using didn't implement the solidgauge chart types from highcharts.js. Will drop the work on solid gauge charts now and try to implement the concret charts for the popup layer: https://github.com/rl-institut/WAM_APP_stemp_abw/issues/22#issuecomment-474297876

Full picture:

I tried to implement the solidgauge charts type (I took as template the highcharts.js example, because there was no example for the python-highcharts wrapper library):

```
class HCSolidGauge(HCStemp):
    setup = {
        'chart': {
            'type': 'solidgauge',
            'backgroundColor': 'rgba(255, 255, 255, 0.0)',
            'height': str(int(9 / 16 * 100)) + '%',
        },
        'pane': {
            'center': ['50%', '85%'],
            'size': '140%',
            'startAngle': -90,
            'endAngle': 90,
            'background': {
                'backgroundColor': 'rgba(255, 255, 255, 0.0)',
                'innerRadius': '60%',
                'outerRadius': '100%',
                'shape': 'arc'
            }
        },
        'tooltip': {
            'enabled': False
        },
        'yAxis': {
            'stops': [
                [0.1, '#55BF3B'],  # green
                [0.5, '#DDDF0D'],  # yellow
                [0.9, '#DF5353']  # red
            ],
            'lineWidth': 0,
            'minorTickInterval': False,
            'tickAmount': 2,
            'title': {
               'y': -70
            },
            'labels': {
                'y': 16
            }
        },
        'plotOptions': {
            'solidgauge': {
                'dataLabels': {
                    'y': 5,
                    'borderWidth': 0,
                    'useHTML': True
                }
            }
        }
    }
```

I had to muddle through different errors messages and solve them one-by-one (by excluding the error prone parts):

- innerRadius Not Allowed For Event ClassOption
- y Not Allowed For Event Class
- Not An Accepted Option Type: solidgauge
- allowed_args = PLOT_OPTION_ALLOWED_ARGS[series_type], KeyError: 'solidgauge'

Finally I realised that the solidgauge charts type is not implemented in the python-highcharts-wrapper library we are using: https://github.com/henhuy/python-highcharts/blob/master/highcharts/highcharts/highchart_types.py#L10-L463

**Comment by nesnoj** [comment]
That's a pity, I'd have liked to use the gauge - we may get back to this later.
Ok, go for the specific charts. Do you think there will be time left today after you'll have finished this?

**Comment by henhuy** [comment]
I had the same problem for some other attributes.
That is the reason why I forked the repo - to implement a workaround and to add needed attributes...
We can figure out later what solution fits our needs best.

**Comment by nesnoj** [comment]
@4lm, I fixed the "2 requests => 2 different HC DIVs" problem in 2a11035a0faf33715cf75f05d6afac0c551794df
It's bloody dirty but works.

If you got some spare time, go ahead ;)

**Comment by nesnoj** [comment]
Some small modifications eb75da12b84f3e8b8c96d25a5e8783011a0cb729. I still don't really like this solution but it's the only way right now.
There were 3 line graphs instead of 1 in the HC, therefore I changed the data but now it's displayed in the wrong dimension - can be easily fixed.

**Comment by 4lm** [comment]
> There were 3 line graphs instead of 1 in the HC, therefore I changed the data but now it's displayed in the wrong dimension - can be easily fixed.

Done: 6fea22ff31f7dc8bb275d56edb290e65daa485a6

**Comment by 4lm** [comment]
Will now (after lunch) add the other charts: https://github.com/rl-institut/WAM_APP_stemp_abw/issues/22#issuecomment-474297876

**Comment by nesnoj** [comment]
Ok, I added more charts above https://github.com/rl-institut/WAM_APP_stemp_abw/issues/22#issuecomment-474297876

**Comment by 4lm** [comment]
Done with all the popup charts from https://github.com/rl-institut/WAM_APP_stemp_abw/issues/22#issuecomment-474297876

**Comment by nesnoj** [comment]
Thanks a lot! :rocket:

**Comment by nesnoj** [comment]
Strangely, in many chart popups there's a legend for entries with little share but none for the one with the largest share, e.g.:
![stemp_app_abw-popup_chart_legend_bug](https://user-images.githubusercontent.com/15910173/54811272-9b85a100-4c88-11e9-8cf9-6752174209f2.png)

Also, the chart could be larger to make it more legible. Maybe the first issue is caused by the small size?

**Comment by 4lm** [comment]
Hi @nesnoj,

yes you are right, the styling of popups and charts are not optimal, as spoken on the phone, I will work on this today:

- Solve the size issue of the charts (in general, make them bigger).
- See if I can get labels in the bigger pie pieces, because I think that the bigger pie pieces should be labeled inside the piece and only smaller pieces should be labeled from outside. I will have a look, why this isn't the case here.
- Maybe we have to think about making popups with charts 2-column wide because the popups will get to high.
- Generalize the styling of the popup, e.g. the height of the charts in the popups is hard-coded in px right now.
- The above point are also connected to #21

We didn't speak about this on the phone, but I also would like to integrate sources and assumptions soonish, because they are also part of the popups and not yet implemented.

**Comment by nesnoj** [comment]
Yepp, sounds good.
BTW, I will ask @Stefanie08 to start integrating the sources by next week

**Comment by nesnoj** [comment]
One more thing: The columns in the column bar chart in layer "Anteil Energie aus EE an Strombedarf" do not show up, this is probably also related to the y-scaling problem discussed above.

**Comment by 4lm** [comment]
So, I pushed 72eae79e1316b4a237c6869ba0489454011fa412 onto the feature/add_more_charts branch.

I tested:

- 2-column popups -> Doesn't look good, because popups look extra-wide, but still have to be relatively wide to fit the pie chart properly and also the flow of looking at the chart is disturbed, because the eyes won't to go either left to right or top to bottom, but at least I have to problem to first go top to bottom on the left side and then go right to look at the chart, I always look at the chart after reading the headline of the popup and then jump back with my eyes - I don't like it that way.

- Then I tried to make the pie chart bigger to fit all data labels next to the pie chart. Result, the pie chart has to be huge which is no solution, also having all labels at once adds quite some visual noise.

- I finally decided to make the pie chart only a little bit bigger and use a legend. And only show the values if the user hovers over a certain pie chart piece. This is less noisy and works IMO. This also solves: https://github.com/rl-institut/WAM_APP_stemp_abw/issues/22#issuecomment-477563681 

One of the next things on my todo list would be to style the popup properly, right there is now real styling ;) But for this, I would like to talk to @bmlancien first!

**Comment by 4lm** [comment]
I will now start to integrate sources and assumptions.

**Comment by 4lm** [comment]
I added sources via an ID list in the layer config files (layers_areas.cfg, layers_region.cfg). It's working, but what I don't like is that the naming is numerical yet (e. g. "1, 2, 3,...") and therefore not descriptive enough. I have to go earlier today. Next week I will implement descriptive naming by taking the names for the sources in the popups directly from the meta-JSON in the database. My first try will be to get the corresponding meta-JSON from the DB and parse it for the title, which I then will use as source link text ...

**Comment by nesnoj** [comment]
I created a RM ticket on the layout for @bmlancien.

Concerning the source list format, you do not have to reinvent the wheel. Just use a similar [template](https://github.com/rl-institut/WAM/blob/master/meta/templates/meta/source_list.html) to the WAM's. @henhuy used jmespath in a [property dec in the source model](https://github.com/rl-institut/WAM/blob/117ac37ddea943fca23a2da74012807bc2a0a2e2/meta/models.py#L45) to parse the json. Just import the source model from WAM into the MasterDetailView and prepare a context based upon a filtered (IDs in config file) queryset.

**Comment by 4lm** [comment]
As spoken, sources can now be added by @Stefanie08 in the database and declared in the config files. If the value in the config file is 0, the sources section is hidden. Right now the backend-user who does do the data entry doesn't have an easy way to look up the database IDs of the sources entries, which have to be entered into the database. Therefore I will overwrite the string representation of the Source class __str__, so that the backend user can easily identify the ID for entering into the config file.

**Comment by 4lm** [comment]
OK, after speaking to @nesnoj I wil not do this. Because @Stefanie08 is doing the sources entries in the database and @nesnoj is doing the sources dispatching in the config files.

**Comment by 4lm** [comment]
Hi @nesnoj,

I will close this issue, which became IMO to overloaded (also a lot of the topics are worked of and already merged to dev) and will open two new ones which are more specific:

- https://github.com/rl-institut/WAM_APP_stemp_abw/tree/feature/style-popup - I will open a new issue on this topic and also want to further talk with @bmlancien how to further improve the popup styling. We already integrated some Foundation classes, so that we can work on the CSS.

- https://github.com/rl-institut/WAM_APP_stemp_abw/tree/feature/refactor-document-popup-related-code - In this branch I want to start to refactor the popup related code and also start to document it.

**Comment by nesnoj** [comment]
@4lm, short question:
You don't use `charts_data.py` or `results.py` in the popups' HC, right?

**Comment by 4lm** [comment]
@nesnoj, right! I used them for testing purposes and they should not be linked with the popups' HC any more! It should be safe to delete `charts_data.py` and rewrite `results.py` to fit your needs ...

**Comment by nesnoj** [comment]
Great, as I will start to refactor the results part now..

**Comment by nesnoj** [comment]
> Finally I realised that the solidgauge charts type is not implemented in the python-highcharts-wrapper library we are using: https://github.com/henhuy/python-highcharts/blob/master/highcharts/highcharts/highchart_types.py#L10-L463

As discussed the the last WAM meeting, it seems that for future apps we will switch to plotly as it can be used for commercial apps (contrary to highcharts) and the integration seems to be way easier (and we got @Bachibouzouk as plotly pro ;) .

@4lm Nevertheless, could you spare some minutes to estimate how long it'd take to add a new chart type to python-highcharts, e.g. the gauge?

**Comment by 4lm** [comment]
> @4lm Nevertheless, could you spare some minutes to estimate how long it'd take to add a new chart type to python-highcharts, e.g. the gauge?

Hi @nesnoj, right now I'm working in REEEM, later this day I will have a closer look at it and will try to give you an estimate ...

**Comment by 4lm** [comment]
A lot of traffic here :wink:, I'm reopening this issue!

**Comment by 4lm** [comment]
Hi @nesnoj,

I had a look at the [python-wrapper for highcharts](https://github.com/henhuy/python-highcharts), it looks generic (so new chart types should be addable to the existing code base without rewriting the whole thing). But the code lacks developer documentation for adding further charts, it only has usage examples.

Right now I can't really guess how long it would take, a wild guess would be 1 to 3 working days, if there are no show stoppers.

My approach would be explorative. I would build the solid gauge chart in pure JS, to see which [JS-API functions](https://api.highcharts.com/highcharts/plotOptions.solidgauge) have to be implemented (because there are a ton for the solidgauge charts) into the Python code base and then try to implement them one after another ...

What to do you think?

**Comment by nesnoj** [comment]
> Right now I can't really guess how long it would take, a wild guess would be 1 to 3 working days, if there are no show stoppers.

Uh, that seems to be a lot of wasted time... let's talk later today
